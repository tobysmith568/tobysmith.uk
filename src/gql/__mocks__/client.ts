export const client = {
  request: jest.fn()
};
export { gql } from "graphql-request";
