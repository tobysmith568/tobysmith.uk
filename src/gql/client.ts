import { GraphQLClient } from "graphql-request";
import { getEnv } from "../utils/api-only/env";

const { apiUrl } = getEnv();

export const client = new GraphQLClient(apiUrl, { headers: {} });
export { gql } from "graphql-request";
export type { Variables as VariablesBase } from "graphql-request";
