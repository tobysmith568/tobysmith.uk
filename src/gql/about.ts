import { client, gql } from "./client";
import Seo from "./seo";

// cSpell:words abouts

export interface About {
  avatar: {
    url: string;
  };
  topText: {
    html: string;
  };
  mainText: {
    html: string;
  };
  tags: {
    name: string;
    url: string;
    icon?: {
      url: string;
    };
  }[];
  seo: Seo;
}

interface GetAboutResponse {
  abouts: About[];
}

const getAbout = async (): Promise<About> => {
  const query = gql`
    query About {
      abouts(first: 1) {
        avatar {
          url
        }
        topText {
          html
        }
        mainText {
          html
        }
        tags {
          name
          url
          icon {
            url
          }
        }
        seo {
          title
          description
          noIndex
        }
      }
    }
  `;

  const { abouts } = await client.request<GetAboutResponse>(query);

  if (abouts.length !== 1) {
    throw new Error(`Expected exactly one about but got ${abouts.length}`);
  }

  return abouts[0];
};
export default getAbout;
