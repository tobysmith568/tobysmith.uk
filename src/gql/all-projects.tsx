import { client, gql } from "./client";
import Seo from "./seo";

export interface AllProjectsPage {
  categories: Category[];
  seo?: Seo;
}

export interface Category {
  name: string;
  items: Item[];
}

export interface Item {
  title: string;
  content: {
    text: string;
  };
  url?: string;
  abandoned?: boolean;
  incomplete?: boolean;
  owner: Owner;
}

type Owner = "Owner" | "JointOwner" | "Contributor";

interface AllProjectsResponse {
  projectIndexPages: AllProjectsPage[];
}

const getAllProjectsPage = async (): Promise<AllProjectsPage> => {
  const query = gql`
    query AllProjects {
      projectIndexPages(first: 1) {
        categories {
          name
          items {
            title
            content {
              text
            }
            url
            abandoned
            incomplete
            owner
          }
        }
        seo {
          title
          description
          noIndex
        }
      }
    }
  `;

  const { projectIndexPages } = await client.request<AllProjectsResponse>(query);

  if (projectIndexPages.length !== 1) {
    throw new Error(`Expected exactly one project page but got ${projectIndexPages.length}`);
  }

  return projectIndexPages[0];
};
export default getAllProjectsPage;
