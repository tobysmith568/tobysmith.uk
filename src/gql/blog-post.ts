import { client, gql } from "./client";
import Seo from "./seo";

export interface BlogPost {
  title: string;
  date: string;
  content: {
    html: string;
  };
  seo?: Seo;
}

export interface BlogPostResponse {
  post?: BlogPost;
}

interface Variables {
  slug: string;
}

const getBlogPost = async (slug: string): Promise<BlogPost> => {
  const query = gql`
    query Post($slug: String) {
      post(where: { slug: $slug }) {
        title
        date
        content {
          html
        }
        seo {
          title
          description
          noIndex
        }
      }
    }
  `;

  const variables: Variables = {
    slug
  };

  const { post } = await client.request<BlogPostResponse, Variables>(query, variables);

  if (!post) {
    throw new Error(`Could not find a blog post with the slug ${slug}`);
  }

  return post;
};
export default getBlogPost;
