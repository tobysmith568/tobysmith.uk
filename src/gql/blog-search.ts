import { client, gql, VariablesBase } from "./client";
import Seo from "./seo";

interface PostPage {
  seo?: Seo;
}

export interface Post {
  slug: string;
  title: string;
  date: string;
  excerpt: string;
}

export interface BlogResponse {
  postPages: PostPage[];
  posts: Post[];
}

export interface Blog {
  seo?: Seo;
  posts: Post[];
}

interface Variables extends VariablesBase {
  term: string;
}

const getBlogSearchPosts = async (term: string): Promise<Blog> => {
  const query = gql`
    query Search($term: String) {
      posts(
        orderBy: date_DESC
        where: { AND: [{ _search: $term }, { OR: [{ devOnly: null }, { devOnly: false }] }] }
      ) {
        slug
        title
        date
        excerpt
      }
      postPages(first: 1) {
        seo {
          title
          description
          noIndex
        }
      }
    }
  `;

  const variables: Variables = {
    term
  };

  const { postPages, posts } = await client.request<BlogResponse, Variables>(query, variables);

  if (postPages.length !== 1) {
    throw new Error(`Expected exactly one postPage but got ${postPages.length}`);
  }

  return { seo: postPages[0].seo, posts };
};
export default getBlogSearchPosts;
