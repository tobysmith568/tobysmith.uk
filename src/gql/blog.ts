import { getEnv } from "../env";
import { client, gql } from "./client";
import Seo from "./seo";

interface PostPage {
  seo?: Seo;
}

export interface Post {
  slug: string;
  title: string;
  date: string;
  excerpt: string;
}

interface BlogResponse {
  postPages: PostPage[];
  posts: Post[];
}

export interface Blog {
  seo?: Seo;
  posts: Post[];
}

const getBlogPosts = async (): Promise<Blog> => {
  const query = gql`
    query Blog {
      posts(orderBy: date_DESC ${allowDevOnly()}) {
        slug
        title
        date
        excerpt
      }
      postPages(first: 1) {
        seo {
          title
          description
          noIndex
        }
      }
    }
  `;

  const { postPages, posts } = await client.request<BlogResponse>(query);

  if (postPages.length !== 1) {
    throw new Error(`Expected exactly one postPage but got ${postPages.length}`);
  }

  return { seo: postPages[0].seo, posts };
};
export default getBlogPosts;

const allowDevOnly = (): string => {
  if (getEnv().isDevelopment) {
    return ", where: {OR: [{devOnly: null}, {devOnly: false}]}";
  }

  return "";
};
