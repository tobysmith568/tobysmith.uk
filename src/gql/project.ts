import highlightCode from "../utils/api-only/highlight-code";
import { client, gql } from "./client";
import Seo from "./seo";

export interface Project {
  title: string;
  subtitle: string;
  content: {
    html: string;
  };
  seo?: Seo;
}

interface ProjectResponse {
  project: Project;
}

interface Variables {
  slug: string;
}

const getProject = async (slug: string): Promise<Project> => {
  const query = gql`
    query GetProject($slug: String) {
      project(where: { slug: $slug }) {
        title
        subtitle
        content {
          html
        }
        seo {
          title
          description
          noIndex
        }
      }
    }
  `;

  const variables: Variables = {
    slug
  };

  const { project } = await client.request<ProjectResponse, Variables>(query, variables);

  if (!project) {
    throw new Error(`Could not find a project with the slug ${slug}`);
  }

  project.content.html = highlightCode(project.content.html);

  return project;
};
export default getProject;
