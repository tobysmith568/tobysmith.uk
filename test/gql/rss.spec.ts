import { client } from "../../src/gql/client";
import getRssFeedData, { RssFeedData } from "../../src/gql/rss";

jest.mock("../../src/gql/client");

describe("gql rss", () => {
  const mockedClient = jest.mocked(client);

  const fakeRssResult: RssFeedData = {
    rss: {
      title: "My RSS Feed"
    },
    posts: [{ title: "Post #1" }, { title: "Post #2" }]
  } as RssFeedData;

  beforeEach(() => {
    jest.resetAllMocks();

    mockedClient.request.mockResolvedValue(fakeRssResult);
  });

  afterAll(() => jest.restoreAllMocks());

  describe("getRssFeedData", () => {
    it("should call request with the correct gql", async () => {
      await getRssFeedData();

      const actualGql = mockedClient.request.mock.calls[0][0];
      expect(actualGql).toMatchSnapshot();
    });

    it("should throw if the client returns undefined", async () => {
      mockedClient.request.mockResolvedValue(undefined);

      await expect(getRssFeedData()).rejects.toThrow("Unable to get RSS data");
    });

    it("should return the rss data when the client returns it", async () => {
      const rssData = await getRssFeedData();

      expect(rssData).toStrictEqual(fakeRssResult);
    });
  });
});
