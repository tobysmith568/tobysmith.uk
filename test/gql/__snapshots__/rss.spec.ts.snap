// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gql rss getRssFeedData should call request with the correct gql 1`] = `
"
    query GetRss {
      rss(where: { type: \\"blog\\" }) {
        title
        description
        siteUrl
        feedUrl
        language
        timeToLive
        webmaster
        copyright
      }
      posts(orderBy: date_DESC, where: { OR: [{ devOnly: null }, { devOnly: false }] }) {
        slug
        title
        date
        excerpt
      }
    }
  "
`;
